addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/ExternalRedirectScanRule.java

```diff
@@ -350,7 +352,27 @@ private int isRedirected(String payload, HttpMessage msg) {
         // Location: http://www.example.org/index.php
         String value = msg.getResponseHeader().getHeader(HttpHeader.LOCATION);
         if (checkPayload(value, payload)) {
-            return REDIRECT_LOCATION_HEADER;
+            URI locUri = null;
+            URI payloadUri = null;
+            try {
+                locUri = new URI(value, true);
+                payload =
+                        StringUtils.startsWithIgnoreCase(REDIRECT_SITE, payload)
+                                ? HttpHeader.SCHEME_HTTP + payload
+                                : payload;
+                payloadUri = new URI(payload, true);
+            } catch (URIException e) {
+                logger.debug(e.getMessage(), e);
+            }
+            try {
+                if (locUri != null
+                        && locUri.getHost() != null
+                        && locUri.getHost().equals(payloadUri.getHost())) {
+                    return REDIRECT_LOCATION_HEADER;
+                }
+            } catch (URIException e) {
+                logger.debug(e.getMessage(), e);
+            }
         }
 
         // (2) Check if redirection by "Refresh" header
```

Comment:
Couldnt the same thing happen in a Refresh header as above, ie the payload in a url param? If so we could have a checkUrlPayload which does the same check for both. 
Thinking about it ... could it happen in all of the cases tested for?
e.g. `<meta http-equiv="location" content="URL=https://safe.com?url=http://evil.com" `

Reply:
I've extracted the method, I still need to implement it further and add more tests.